Design-Docs/Cluster-GitOps.md

```diff
@@ -0,0 +1,78 @@
+# Proposal for GitOps Style IaC for Cluster Management
+
+Kurt Garloff, v0.1, 2022-02-18
+
+## Motivation and Goals
+
+Using the Kubernetes Cluster-API (capi), we can use a k8s style declarative
+way to describe the workload clusters that should be running and can manage
+their lifecycle: creation, changes, rolling upgrades and clean up can all be
+performed with it. The OpenStack provider (capo) has the basic integration
+to manage the networks, virtual machines, load-balancers. For full automation,
+a few more pieces have been developed in the SCS
+[k8s-cluster-api-provider](https://github.com/SovereignCloudStack/k8s-cluster-api-provider/)
+repository: Registering the images, setting up extra security groups for
+non-Calico CNI, creating anti-affinity server groups, deploying the OCCM
+and cinder CSI integration and optionally some more services (such as Flux,
+cert-manager, nginx ingress, ...)
+
+This follows similar ideas as described on
+<https://www.weave.works/blog/gitops-and-cluster-api-master-of-masters>.
+
+Most of the simpler cluster setups can be done without ever touching the cluster-template.yaml
+file -- just doing a dozen adjustments in clusterctl.yaml provides a reasonable amount
+of flexibility. In SCS' k8s-cluster-api-provider setup, the standard settings from
+the capo templates have been extended by the settings that let you chose cilium
+as alternative CNI provider, anti-affinity, the OCCM and CSI deployment and the extra services.
+
+To implement a simple gitops style management for a set of clusters, we would
+basically create a reconciliation loop on the capi management node, which
+gets the enhanced configuration files from git and creates, changes and
+destroys clusters according to the settings there. The mechanism should
+distinguish between base settings and overlays (kustomization style) to
+modify and extend the settings and should allow an opt-in mechanism to
+do more detailed adjustments via overriding also the cluster-templates
+(kustomization style again).
+
+Cluster-admin credentials need to be provided to the owners of the cluster.
+The current thinking is to have a public key included in the git repo;
+the cluster-admin credentials for the created cluster would be encrypted
+with this public key and can then be published. (Would still use https
+to ensure that these are the true credentials.) Only the owner(s) of
+the private key can decrypt the credentials.
+
+## Implementation thoughts
+
+The reconciliation loop would roughly look like this:
+1. Get the latest clusters from git (via a regular check or an event)
+1. Per cluster
```

Comment:
I'm not sure what (and why) exactly is needed here beyond what already happens when one applies a cluster resource.

In my mind, the workflow would look like: "install fluxcd, register git repository, put cluster manifests in there, done", so this looks very complex. Am I missing anything?